### ISSUE TITLE
ntsOId and xNId

### ISSUE THREAD

**Issue by emiltin** [rsmp-nordic/rsmp_core#9]
Messages include the fields ntsOId and xNId. They seem to refer to a swedish national system of ids? From https://github.com/rsmp-nordic/rsmp_core/blob/master/rst/rsmp.rst:

- ntsOId | NTSObjectId | Component id for the NTS object which the message is referring to.
- xNId | externalNtsId | Identity for the NTS object in communcation between NTS and other systems. The format is 5 integers. Defined in cooperation with representatives from NTS. Unique for the site.

To ensure RSMP is useful as a regional/international standard, it would be best to avoid attributes that are particular to sweden or any other single country.

**Comment by otterdahl** [comment]
I agree.
Originally it seemed like a good idea to have these in the protocol, but they never got the used the way that it was originally intended. However, there need to be some discussions with other technical areas (beyond traffic signal)s to make sure that no one else is using them.

**Comment by emiltin** [comment]
We depreciate these two fields. Will be removed in a later release (not 3.1.5)

**Comment by otterdahl** [comment]
The relation between ntsOId and componendId is used to logically group **grouped objects** and **single objects** together. For instance, traffic signals has **Traffic controller** as a grouped object with each associated signal group and detector logic as single objects.

This needs to be resolved before removing these fields

**Comment by emiltin** [comment]
I don't understand the relation between ntsOId and componendId, that you mentioned above.

**Comment by otterdahl** [comment]
On second thought, they might not be needed after all.

The relation between which component (e.g. signal group) that belongs to which main object (e.g. Traffic controller) isn't really important from the protocol point of view.

The reason for my hesitation is due to the fact that RSMP supports multiple main components in the same SXL (and optionally multiple site-id). We don't use multiple main components in the SXL for Traffic signals, but it is used in other SXL's, such as congestion tax or intelligent street lighting.

That means that it is technically possible to send alarms, commands and statuses for several traffic controllers in the same RSMP connection. This is needed in order to support infrastructure-to-infrastructure (I2I)

**Comment by emiltin** [comment]
So can the ntsOId and xNId attributes be deprecated from 3.1.6 or is this still pending further investigations?

**Comment by otterdahl** [comment]
Pending further investigation. I'm inviting people responsible for VMS-signs in upcoming working group meetings to discuss this.

**Comment by emiltin** [comment]
ok

**Comment by emiltin** [comment]
There has been some recent discussion in https://github.com/rsmp-nordic/rsmp_validator/issues/59.

Reading core spec, the ntsOId and xNId are described under Basic Message Structure, under 'variable content':
https://rsmp-nordic.org/rsmp_specifications/core/3.1.5/applicability/basic_structure.html#basic-structure

It states that the content is defined in the SXL. What does that mean? Looking in the SXL for TCL's I see no explanation of these two fields.

Another question is related to whether these fields are optional or required. It seems TLCs expect xNId to be an empty string "" and returns errors if null is used, or the attribute is left out. Most (all?) json examples also use the empty string "" for the nXId attribute. If we can clarify this we could also update our JSON schema. Currently the scheme does required neither ntsOId nor xNId.

Back in 2020 we wrote that depreciating the ntsOId and xNId attributes was pending further investigation. Is there news on this, or is it something we should pick up again?

In any case I think it would be nice to expand the explanation about the fields in the core spec. These fields have confused me quite a bit.

**Comment by emiltin** [comment]
since this is a breaking change (though a smaller one) we will push this to 4.0

**Comment by emiltin** [comment]
Until these confusing ntsOId and xNId attributes are removed, can we at least allow them to be set to null or an empty string? I don't see any reason to have to duplicate the component id in ntsOId for TLCs.

I see some controllers complaning if they are missing, but others don't.

**Comment by emiltin** [comment]
> The reason for my hesitation is due to the fact that RSMP supports multiple main components in the same SXL (and optionally multiple site-id). We don't use multiple main components in the SXL for Traffic signals, but it is used in other SXL's, such as congestion tax or intelligent street lighting.

In these cases, the RSMP connection (and the equipment handling it) act as a gateway, distributing messages to their final receiver?

Is there a need for more robust routing of messages?

IP addresses and domain names is a classical example of routing. RSMP already relies on this in the sense that sites connects to supervisors using ip address or domain name.

**Comment by otterdahl** [comment]
> can we at least allow them to be set to null? I don't see any reason to have to duplicate the component id in ntsOId.

Absolutely. Perhaps we should add a clarification in core, just like we did with functionalPosition and functionalState.

**Comment by otterdahl** [comment]
> In these cases, the RSMP connection (and the euipment handling it) act as a gateway, distributing messages to their final receiver?

Yes, it kinda allows you to use RSMP between supervision systems. Or just use a single RSMP connection to several controllers.

> Is there a need for more robust routing of messages?
> 
> IP addresses and domain names is a classical example of routing. RSMP relies on this since each sites connects to the supervisor using ip address or domain name.

I think that ntsOId is part of the RSMP legacy and once we adopt MQTT or similar this support is outdated. I'm favor of keeping it simple and use a separate connection for each TLC instead.

**Comment by emiltin** [comment]
> can we at least allow them to be set to null? I don't see any reason to have to duplicate the component id in ntsOId.
>> Absolutely. Perhaps we should add a clarification in core, just like we did with functionalPosition and functionalState.

Should we allow ntsOId and xNId to be left out? I think that's very similar to using "", but a bit cleaner.

I guess it's debatable whether that's a breaking change, although both essentially means the same: the fields are not used.

**Comment by otterdahl** [comment]
I'm in favor of removing ntsOId and xNId completely. I'm convinced that nobody are using them. Can we do this in 3.3.0 or should we clarify that they should be set to null in order to avoid a breaking change? @emiltin

**Comment by emiltin** [comment]
Unless the current core spec allows you to leave these fields out, I don't think we can remove then in a minor 3.3.0 version. But we can state that they are depreciated and should be to set empty strings by senders, and ignored by the receivers.

We could perhaps add some general flexibility by stating in the core spec that if a field can be set to nil or an empty string, it's valid to leave out the field as an alternative. That would require implementations to handle optional fields, meaning removing a field should not break exisiting implementations, as they woudl jsut interpret is as nil or an empty string, depending on the field?

**Comment by emiltin** [comment]
Currently we require all connections to use exactly matching core versions. So in practice backward compatibility is not relevant when it comes to communication, but only when it comes to implementations, i.e. do you need to change your code.

This is an area that I would like us to be more clear on. Are we also aiming for site/supervisors with different minor core versions to communicate or not? #100
